src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs

```diff
@@ -0,0 +1,184 @@
+                case SymbolKind.TypeParameter:
+                    publicIcon = Glyph.TypeParameter;
+                    break;
+                default:
+                    throw new ArgumentException();
+            }
+
+            switch (declaredAccessibility)
+            {
+                case Accessibility.Private:
+                    publicIcon += Glyph.ClassPrivate - Glyph.ClassPublic;
+                    break;
+
+                case Accessibility.Protected:
+                case Accessibility.ProtectedAndInternal:
+                case Accessibility.ProtectedOrInternal:
+                    publicIcon += Glyph.ClassProtected - Glyph.ClassPublic;
+                    break;
+
+                case Accessibility.Internal:
+                    publicIcon += Glyph.ClassInternal - Glyph.ClassPublic;
+                    break;
+            }
+
+            return publicIcon;
+        }
+
+        private void AddBuilder(CompletionContext completionContext)
+        {
+            completionContext.SuggestionModeItem = CommonCompletionItem.Create(CSharpFeaturesResources.Name, CompletionItemRules.Default);
+        }
+
+        private ITypeSymbol UnwrapType(ITypeSymbol type)
+        {
+            var nts = type as INamedTypeSymbol;
+            if (nts != null && nts.ConstructedFrom != null)
+            {
+                var constructedFrom = nts.ConstructedFrom;
+                switch (constructedFrom.SpecialType)
+                {
+                    case SpecialType.System_Collections_Generic_IEnumerable_T:
+                    case SpecialType.System_Collections_Generic_IList_T:
+                    case SpecialType.System_Collections_Generic_ICollection_T:
+                    case SpecialType.System_Collections_Generic_IReadOnlyList_T:
+                    case SpecialType.System_Collections_Generic_IReadOnlyCollection_T:
+                    case SpecialType.System_Nullable_T:
+                        return UnwrapType(nts.TypeArguments[0]);
+                }
+
+                if (constructedFrom.Name == "Task"
+                   && constructedFrom.ContainingNamespace?.Name == "Tasks"
+                   && constructedFrom.ContainingNamespace.ContainingNamespace?.Name == "Threading"
+                   && constructedFrom.ContainingNamespace.ContainingNamespace.ContainingNamespace?.Name == "System"
+                   && (constructedFrom.ContainingNamespace.ContainingNamespace.ContainingNamespace.ContainingNamespace?.IsGlobalNamespace ?? false))
+                {
+                    return UnwrapType(nts.TypeArguments[0]);
+                }
+            }
+
+            return type;
+        }
+
+        private IEnumerable<(string, SymbolKind)> GetRecommendedNames(IEnumerable<Words> baseNames, NameDeclarationInfo declarationInfo, CSharpSyntaxContext context)
+        {
+            var namingStyleOptions = context.Workspace.Options.GetOption(SimplificationOptions.NamingPreferences, LanguageNames.CSharp);
```

Comment:
are naming preferences not part of editor configu?  Shouldn't we get these for the document we're in?